### ISSUE TITLE
[Question] Way to read token from url

### ISSUE THREAD

**Issue by airguillaume** [thephpleague/oauth2-server#1016]
Hi there : ),

I am currently using Slim v3 and this package. I successfully integrated the implementation for my API routes.

However, in a confirmation email I am using a link like this (see below) and I would like to know if there is a way to extract the token from the URL.: 

> http://site.com/confirm/{{access_token_with_specific_scope}}

If this isn't the place to ask, I apologize.

Best regards,
Guillaume R.

**Comment by sg3s** [comment]
If I understand correctly you are sending this email for some kind of account activation?
In any case, getting the access token from a URL is outside of the scope of this library.

I can advice you to do the following things:
1. If you are going to include an access token in a URL, do it in a query parameter, not the path section. The path section is often logged by servers, while query parameters are less often. Even better would be to include it in the hash part of the URL, which is never send to the server, but you will only be able to access the hash on the frontend / in javascript (which you could use to then make another request with the token in a HTTP Header instead).

2. For extracting any data from a request in Slim I would advise to use a middleware specific to your application which knows where a token should be present. You can find examples on how to do this in the official documentation, also describing how to pass variables;
http://www.slimframework.com/docs/v3/concepts/middleware.html

3.a If the primary function of your endpoint is to validate/do something with the access token then you can write the logic to extract the access token in your controller itself. This is assuming you take the advice to move the access token to a query parameter or let javascript do a separate request to the server.

3.b If you keep your URL structure as given you should extract the parameter from the path when defining the route for your controller / handler. This is explained in the "First Application" tutorial;
https://www.slimframework.com/docs/v3/tutorial/first-app.html#create-the-application
https://www.slimframework.com/docs/v3/objects/router.html#how-to-create-routes

A route for you would look something like this;
```
$app->get('/confirm/{token}', function (Request $request, Response $response, array $args) {
    $name = $args['token'];
    $response->getBody()->write("Your token is, $token");
    return $response;
});
```

Again, I would advise against keeping the token in the path part of your URL; here is some more specific documentation on how you can transfer variables between middleware and controller;
https://www.slimframework.com/docs/v3/objects/request.html#attributes

**Comment by airguillaume** [comment]
Hi, sorry for the late response.

Ok so, I think I might have wrongly explained my problem. It isn't how to get the token but how to read it from a string and authorize a request. Is there a way do to something like this ?

```
$app->get('/confirm/', function (Request $request, Response $response, array $args) {
    $token = $args['token'];

    // Something here to do to authorize the request e.g :
    // if (!$app->authorizeFromString($token)) {
        // return false;
    // }

    $response->getBody()->write("authorized");
    return $response;
});
```

Thanks for the response : )

**Comment by sg3s** [comment]
That does make more sense. I think I know what you're after. You want to know how to validate the access token in a non-standard situation.

The library supplies a middleware for the standard situation where a bearer token is validated;
https://oauth2.thephpleague.com/resource-server/securing-your-api/

To validate the token in a request in a non-standard way, I would simply add a middleware which modifies the request to include the token in the Authorization header as a bearer token, before this supplied middleware processes the request.

This could be as easy as;

    return $next($request->withHeader('Authorization', "Bearer $token"), $response);

Though you would need to retrieve `$token` from the request before this, and you might want to make modifications to the path and catch requests lacking a token in the correct position maybe.

Doing that you should be able to use the supplied middleware to validate the request. If this is not an option I would suggest writing your own middleware/jwt validator. There are multiple packages available to make it easier.

**Comment by Sephster** [comment]
Thanks for your responses with this @sg3s - much appreciated for your assistance. I think @sg3s has covered most of what you will need. I would also suggest you look at the `BearerTokenValidator` class to find out how the OAuth server validates a token, specifically the `validateAuthorization()` function.

Most of the validation is handled by our third party JWT library. We use lcobucci/jwt.

**Comment by airguillaume** [comment]
Thank you so much for your responses. This helped me a lot.
 I did not tought about this simple solution that is setting the token in the header after getting it from the url. This works well.

Have a good day :) .

**Comment by airguillaume** [comment]
Hey, it's me again ! 

I'm not sure if I can ask on top of this question (if not, I still can create a new issue/question ^^) but; I'm trying to manually generate an access token. I followed the methods issueAccessToken, getNewToken and tried to implement it but without success. Is there a link, a resource for this ? I tried a research in the issues but couldn't find anything relevant. 

Any help would be appreciated :).
